src/dft/src/ScanReplace.cpp

```diff
@@ -0,0 +1,426 @@
+    const sta::LibertyPgPort* non_scan_cell_port,
+    const sta::LibertyCell* scan_cell)
+{
+  sta::LibertyCellPgPortIterator scan_cell_ports_iter(scan_cell);
+  while (scan_cell_ports_iter.hasNext()) {
+    sta::LibertyPgPort* scan_cell_port = scan_cell_ports_iter.next();
+    const bool port_equiv
+        = sta::LibertyPgPort::equiv(non_scan_cell_port, scan_cell_port);
+    if (port_equiv) {
+      return scan_cell_port;
+    }
+  }
+  return nullptr;
+}
+
+// We look for each of the non_scan_cell's ports for a functional equivalent one
+// in the scan_cell. If there is a port that we are unable to find an scan
+// equivalent, then we can say that the given scan cell is not a replacement for
+// the given non-scan one
+bool IsScanEquivalent(
+    const sta::LibertyCell* non_scan_cell,
+    const sta::LibertyCell* scan_cell,
+    std::unordered_map<std::string, std::string>& port_mapping)
+{
+  sta::LibertyCellPortIterator non_scan_cell_ports_iter(non_scan_cell);
+  while (non_scan_cell_ports_iter.hasNext()) {
+    sta::LibertyPort* non_scan_cell_port = non_scan_cell_ports_iter.next();
+    sta::LibertyPort* scan_equiv_port
+        = FindEquivalentPortInScanCell(non_scan_cell_port, scan_cell);
+    if (!scan_equiv_port) {
+      return false;
+    } else {
+      port_mapping.insert(
+          {non_scan_cell_port->name(), scan_equiv_port->name()});
+    }
+  }
+
+  sta::LibertyCellPgPortIterator non_scan_cell_pg_ports_iter(non_scan_cell);
+  while (non_scan_cell_pg_ports_iter.hasNext()) {
+    sta::LibertyPgPort* non_scan_cell_pg_port
+        = non_scan_cell_pg_ports_iter.next();
+    sta::LibertyPgPort* scan_equiv_port
+        = FindEquivalentPortInScanCell(non_scan_cell_pg_port, scan_cell);
+    if (!scan_equiv_port) {
+      return false;
+    } else {
+      port_mapping.insert(
+          {non_scan_cell_pg_port->name(), scan_equiv_port->name()});
+    }
+  }
+
+  return true;
+}
+
+// We calculate the difference in performance of the ports given
+double DifferencePerformancePorts(const sta::LibertyPort* port1,
+                                  const sta::LibertyPort* port2)
+{
+  double diff = 0;
+
+  diff += std::abs(port1->capacitance() - port2->capacitance());
+  diff += std::abs(port1->driveResistance() - port2->driveResistance());
+
+  bool exist_fanout_load1 = false, exist_fanout_load2 = false;
+  float fanout_load1 = 0, fanout_load2 = 0;
+
+  port1->fanoutLoad(fanout_load1, exist_fanout_load1);
+  port2->fanoutLoad(fanout_load2, exist_fanout_load2);
+
+  if (exist_fanout_load1 && exist_fanout_load2) {
+    diff += std::abs(fanout_load1 - fanout_load2);
+  }
+
+  return diff;
+}
+
+// We calculate the difference in performance between the cells iterating
+// through the ports
+double DifferencePerformanceCells(const sta::LibertyCell* non_scan_cell,
+                                  const sta::LibertyCell* scan_cell)
+{
+  double diff = 0;
+
+  // Area
+  diff += std::abs(non_scan_cell->area() - scan_cell->area());
```

Comment:
You are adding area to leakage to delay which makes no sense.  They are totally different units.

Reply:
You are right. I wonder if you have a better criteria in mind to select the best scan cell for a given cell. One alternative is to use the edit distance of the names of the cell vs the scan cell.

Usually in a lib, the scan cell equivalent will have a similar name to the non-scan one like:

- sky130_fd_sc_hd__dfstp_1
- sky130_fd_sc_hd__sdfsbp_1

Of course, this is just an heuristic and maybe there is a better way to select it. What do you think?

Reply:
Here's an alternate way of measuring similarity that preserves units.

Treat attributes as vectors example, a = <area, leakage, port speed ,etc.>. Then norm the vector a/|a|, multiply vector by 1 over the magnitude.

Once you have a list of normed vectors for all the scan flops use the [Cosine Similarity]( https://en.wikipedia.org/wiki/Cosine_similarity) which ranges from 0 -> 1 where 1 is identical. Select max value.

Reply:
The elements of the vector are in very different scales so norming the vector won't improve that.  I think finding the nearest output strength is probably enough.

Reply:
You can subtract the mean vector to normalize scale.

Reply:
Even then it doesn't make sense to try to match area and leakage. Scan cells are going to be larger than their equivalent non scan versions. Output drive strength will matter to preserving timing

Reply:
Fair enough on leakage equivalent, but isn't the resizer going to change these cells anyways? It seems like the things that's most important is D -> Q latency

Reply:
There is no d->q path on an ff, I guess you mean clk->q.  I think the insertion should try to preserve the current timing regardless of what optimizations will happen later on.  The output drive strength is the main determiner of that.  Roughly you want an X2 gate to stay an X2 gate after swapping.

Reply:
Thanks @QuantamHD and @maliberty for the input here!

Matt, what is the current way of getting the output drive for a given LibertyPort? Thanks!

Reply:
LibertyPort::driveResistance() should work

Reply:
Thanks!, I just changed it.